lib/rule_engine/ast.py

```diff
@@ -372,27 +386,78 @@ def to_graphviz(self, digraph, *args, **kwargs):
 
 class AddExpression(LeftOperatorRightExpressionBase):
 	"""A class for representing addition expressions from the grammar text."""
-	compatible_types = (DataType.FLOAT,DataType.STRING)
+	compatible_types = (DataType.FLOAT, DataType.STRING, DataType.DATETIME, DataType.TIMEDELTA)
 	result_type = DataType.UNDEFINED
 
 	def __init__(self, *args, **kwargs):
 		super(AddExpression, self).__init__(*args, **kwargs)
 		if self.left.result_type != DataType.UNDEFINED and self.right.result_type != DataType.UNDEFINED:
-			if self.left.result_type != self.right.result_type:
+			if self.left.result_type == DataType.DATETIME:
+				if self.right.result_type != DataType.TIMEDELTA:
+					raise errors.EvaluationError('data type mismatch')
+				self.result_type = self.left.result_type
```

Comment:
Just a note to myself that left must be a datetime while right is a timedelta and not the other way otherwise Python will throw an error.

<details>
<summary>Python Test</summary>

```
In [1]: import datetime

In [2]: td = datetime.timedelta(seconds=30)

In [3]: td - datetime.datetime.now()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 td - datetime.datetime.now()

TypeError: unsupported operand type(s) for -: 'datetime.timedelta' and 'datetime.datetime'

In [4]: datetime.datetime.now() - td
Out[4]: datetime.datetime(2022, 7, 11, 19, 30, 51, 992026)

In [5]: 
```
</details>

Rule Engine could hypothetically detect this and swap the values but I'm not sure that's the right thing to do so this seems good for now. 👍🏻  Maybe it should just be called out in the docs somewhere (if it's not already).

Reply:
I think it wouldn’t make sense semantically even if it was viable.

And yeah, I’m pretty sure I did outline the possible operations in the documentation updates.